test/consumer/message_tracker_test.exs

```diff
@@ -0,0 +1,64 @@
+defmodule Mississippi.Consumer.MessageTracker.Test do
+  use ExUnit.Case
+
+  alias Mississippi.Consumer.MessageTracker
+
+  require Logger
+
+  setup_all do
+    start_supervised!({Registry, [keys: :unique, name: Registry.MessageTracker]})
+
+    start_supervised!(
+      {DynamicSupervisor, strategy: :one_for_one, name: MessageTracker.Supervisor}
+    )
+
+    :ok
+  end
+
+  doctest Mississippi.Consumer.MessageTracker
+
+  test "a MessageTracker process is successfully started with a given sharding key" do
+    sharding_key = "sharding_#{System.unique_integer()}"
+    {:ok, pid} = MessageTracker.get_message_tracker(sharding_key)
+
+    mt_processes = Registry.select(Registry.MessageTracker, [{{:"$1", :_, :_}, [], [:"$1"]}])
+
+    assert [{:sharding_key, ^sharding_key}] = mt_processes
+
+    on_exit(fn -> DynamicSupervisor.terminate_child(MessageTracker.Supervisor, pid) end)
+  end
+
+  test "a MessageTracker process is not duplicated when using the same sharding key" do
+    sharding_key = "sharding_#{System.unique_integer()}"
+    {:ok, pid} = MessageTracker.get_message_tracker(sharding_key)
+
+    mt_processes = Registry.select(Registry.MessageTracker, [{{:"$1", :_, :_}, [], [:"$1"]}])
+    assert [{:sharding_key, ^sharding_key}] = mt_processes
+
+    {:ok, ^pid} = MessageTracker.get_message_tracker(sharding_key)
+    assert [{:sharding_key, ^sharding_key}] = mt_processes
+
+    on_exit(fn -> DynamicSupervisor.terminate_child(MessageTracker.Supervisor, pid) end)
+  end
+
+  test "a MessageTracker process is spawned again if requested after termination" do
+    sharding_key = "sharding_#{System.unique_integer()}"
+    {:ok, first_pid} = MessageTracker.get_message_tracker(sharding_key)
+
+    mt_processes = Registry.select(Registry.MessageTracker, [{{:"$1", :_, :_}, [], [:"$1"]}])
+    assert [{:sharding_key, ^sharding_key}] = mt_processes
+
+    DynamicSupervisor.terminate_child(MessageTracker.Supervisor, first_pid)
+    # Give some time to the Registry to delete the entry
+    Process.sleep(:timer.seconds(2))
+    mt_processes = Registry.select(Registry.MessageTracker, [{{:"$1", :_, :_}, [], [:"$1"]}])
+    assert [] = mt_processes
+
+    {:ok, second_pid} = MessageTracker.get_message_tracker(sharding_key)
+    mt_processes = Registry.select(Registry.MessageTracker, [{{:"$1", :_, :_}, [], [:"$1"]}])
+    assert [{:sharding_key, ^sharding_key}] = mt_processes
```

Comment:
this is testing that the registry works as intended by removing its entry when it crashes

i suggest we omit this check (maybe adding a comment explaining why?) and focus on checking that the new sharding key is contained in the new `mt_processes`. that way we can avoid the sleep

Reply:
actually we would need to change the `Registry.select` spec so that it also gives us the pid, to make sure the new entry was inserted (as the sharding_key alone does not give us any distinction from the previous entry)

Reply:
The sharding key is the same both bfore and after killing the MessageTracker process. The point of the test is checking that a MessageTracker is spawned and registered correctly by `get_message_tracker/1` even after crashes (or forced termination)